### ISSUE TITLE
Einsum produces transposed output for some cases

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#161729]
```
import torch
import torch.nn.functional as F
import numpy as np

torch.set_default_device("cuda")


torch.set_default_device("cuda")
batch, in_dim, out_dim = 128, 1024, 4096
x = torch.randn(batch, in_dim, dtype=torch.float)
w = torch.randn(out_dim, in_dim, dtype=torch.float)

def linear(x, w):
    return F.linear(x, w)

out = linear(x, w)
print(out.shape, out.stride())
out = torch.einsum("fd,bd->bf", w, x)
print(out.shape, out.stride())
out_np = np.einsum("fd,bd->bf", w.cpu().numpy(), x.cpu().numpy())
print(out_np.shape, out_np.strides)
```
outputs 
```
torch.Size([128, 4096]) (4096, 1)
torch.Size([128, 4096]) (1, 128)
(128, 4096) (16384, 4)
```
`linear` and `numpy` produce contiguous output, `einsum` produces transposed output, there are probably other cases when this happens

cc @jianyuh @nikitaved @mruberry @walterddr @xwang233 @Lezcano

**Comment by ngimel** [comment]
`torch.einsum("bd,fd->bf", x, w)` produces contiguous output (order of args is switched and matches dimension order, now first arg provides first dim of output, second arg - second
